File: HumanJavaApp/app/src/main/java/com/biodigital/humansdksampleapp/HumanActivity.java
package com.biodigital.humansdksampleapp;

import android.animation.LayoutTransition;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.LightingColorFilter;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.ScrollView;

import com.biodigital.humansdk.*;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import static android.icu.lang.UProperty.MATH;

public class HumanActivity extends AppCompatActivity implements HKHumanInterface {

    private HKHuman body;

    private boolean xraymode = false;
    private boolean isolatemode = false;
    private boolean dissectmode = false;
    private boolean paintmode = false;

    ViewPager chapterPager;
    boolean expanded = false;
    HKColor paintColor = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_human);

        body = findViewById(R.id.humanbody);
        body.setInterface(this);
        HashMap<HumanUIOptions,Boolean> uimap = new HashMap<>();
        uimap.put(HumanUIOptions.animation,false);
        body.setUIoptions(uimap);

        String modelID = getIntent().getStringExtra(MainActivity.MODEL_MESSAGE);
        body.load(modelID);
        body.annotations.hide();

        final Button homebutton = (Button)findViewById(R.id.homebutton);
        final Button resetbutton = (Button)findViewById(R.id.resetbutton);
        final Button dissectbutton = (Button)findViewById(R.id.dissectbutton);
        final Button undobutton = (Button)findViewById(R.id.undobutton);
        final Button xraybutton = (Button)findViewById(R.id.xraybutton);
        final Button isolatebutton = (Button)findViewById(R.id.isolatebutton);
        final Button shareButton = (Button)findViewById(R.id.sharebutton);
        final Button paintbutton = (Button)findViewById(R.id.paintbutton);
        final Button redbutton = (Button)findViewById(R.id.redbutton);
        final Button greenbutton = (Button)findViewById(R.id.greenbutton);
        final Button bluebutton = (Button)findViewById(R.id.bluebutton);
        final Button yellowbutton = (Button)findViewById(R.id.yellowbutton);
        final Button undopaintbutton = (Button)findViewById(R.id.undopaintbutton);
        final View allpaintstuff = (View)findViewById(R.id.allpaintstuff);
        final View paintmenu = (View)findViewById(R.id.paintmenu);
        final ScrollView scroller = (ScrollView)findViewById(R.id.scrollView);
        chapterPager = (ViewPager)findViewById(R.id.humanChapterPager);

        homebutton.setVisibility(View.INVISIBLE);
        resetbutton.setVisibility(View.INVISIBLE);
        dissectbutton.setVisibility(View.INVISIBLE);
        undobutton.setVisibility(View.INVISIBLE);
        xraybutton.setVisibility(View.INVISIBLE);
        isolatebutton.setVisibility(View.INVISIBLE);
        shareButton.setVisibility(View.INVISIBLE);
        allpaintstuff.setVisibility(View.INVISIBLE);
        paintmenu.setVisibility(View.INVISIBLE);
        chapterPager.setVisibility(View.INVISIBLE);

        final HKColor redColor = new HKColor();
        final HKColor greenColor = new HKColor();
        final HKColor blueColor = new HKColor();
        final HKColor yellowColor = new HKColor();

        ProgressBar progress = findViewById(R.id.progressBar1);
        progress.setVisibility(View.VISIBLE);

        homebutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                finish();
            }
        });

        resetbutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                body.scene.reset();
                xraymode = false;
                isolatemode = false;
                dissectmode = false;
                xraybutton.getBackground().setColorFilter(null);
                dissectbutton.getBackground().setColorFilter(null);
                isolatebutton.getBackground().setColorFilter(null);
                undobutton.setVisibility(View.INVISIBLE);
            }
        });

        dissectbutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                if (paintmode) {
                    doPaintButton();
                }
                doDissectButton();
            }
        });

        undobutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                body.scene.undo();
            }
        });

        xraybutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                xraymode = !xraymode;
                body.scene.xray(xraymode);
                if (xraymode) {
                    xraybutton.getBackground().setColorFilter(new LightingColorFilter(0xFFFFFFFF, 0xFFAA0000));
                } else {
                    xraybutton.getBackground().setColorFilter(null);
                }
                if ( dissectmode ) {
                    body.scene.dissect(true);
                }
            }
        });

        isolatebutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                isolatemode = !isolatemode;
                body.scene.isolate(isolatemode);
//                if (isolatemode) {
//                    isolatebutton.getBackground().setColorFilter(new LightingColorFilter(0xFFFFFFFF, 0xFFAA0000));
//                } else {
//                    isolatebutton.getBackground().setColorFilter(null);
//                }
                if ( dissectmode ) {
                    body.scene.dissect(true);
                }

            }
        });

        shareButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                body.scene.share();
            }
        });

        paintbutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                if (dissectmode) {
                    doDissectButton();
                }
                doPaintButton();
            }
        });

        redColor.tint[0] = 1.0f;
        redColor.tint[1] = 0.0f;
        redColor.tint[2] = 0.0f;
        greenColor.tint[0] = 0.0f;
        greenColor.tint[1] = 1.0f;
        greenColor.tint[2] = 0.0f;
        blueColor.tint[0] = 0.0f;
        blueColor.tint[1] = 0.0f;
        blueColor.tint[2] = 1.0f;
        yellowColor.tint[0] = 1.0f;
        yellowColor.tint[1] = 1.0f;
        yellowColor.tint[2] = 0.0f;
        greenColor.saturation = 0.5f;
        blueColor.opacity = 0.66f;

        redbutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                paintColor = redColor;
                paintmenu.setBackgroundColor(Color.RED);
            }
        });

        greenbutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                paintColor = greenColor;
                paintmenu.setBackgroundColor(Color.GREEN);
            }
        });

        bluebutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                paintColor = blueColor;
                paintmenu.setBackgroundColor(Color.BLUE);
            }
        });

        yellowbutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                paintColor = yellowColor;
                paintmenu.setBackgroundColor(Color.YELLOW);
            }
        });

        undopaintbutton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                paintColor = null;
                paintmenu.setBackgroundColor(Color.TRANSPARENT);
            }
        });

        chapterPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                if ( position == body.timeline.currentChapter.index ) {
                    System.out.println("already at chapter " + position);
                    return;
                }
                if ( position > body.timeline.currentChapter.index ) {
                    body.timeline.nextChapter();
                } else {
                    body.timeline.prevChapter();
                }
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
    }

    public void handleChapterClick() {
        final float scale = getApplicationContext().getResources().getDisplayMetrics().density;
        RelativeLayout rl = (RelativeLayout) findViewById(R.id.category);
        ((ViewGroup) rl).getLayoutTransition().enableTransitionType(LayoutTransition.CHANGING);
        if (expanded) {
            System.out.println("shrink");
            int px = (int) (50 * scale + 0.5f);
            rl.getLayoutParams().height = px;
        } else {
            System.out.println("expand");
            int px = (int) (160 * scale + 0.5f);
            rl.getLayoutParams().height = px;
        }
        rl.requestLayout();
        expanded = !expanded;
    }
    void doPaintButton() {
        Button paintbutton = (Button)findViewById(R.id.paintbutton);
        View paintmenu = (View)findViewById(R.id.paintmenu);
        paintmode = !paintmode;
        if (paintmenu.getVisibility() == View.VISIBLE) {
            paintmenu.setVisibility(View.INVISIBLE);
            body.scene.enableHighlight();
        } else {
            paintmenu.setVisibility(View.VISIBLE);
            body.scene.disableHighlight();
        }
        if (paintmode) {
            paintbutton.getBackground().setColorFilter(new LightingColorFilter(0xFFFFFFFF, 0xFFAA0000));
        } else {
            paintbutton.getBackground().setColorFilter(null);
        }
    }

    void doDissectButton() {
        Button dissectbutton = (Button)findViewById(R.id.dissectbutton);
        Button undobutton = (Button)findViewById(R.id.undobutton);
        dissectmode = !dissectmode;
        body.scene.dissect(dissectmode);
        dissectbutton.setSelected(dissectmode);
        if (dissectmode) {
            dissectbutton.getBackground().setColorFilter(new LightingColorFilter(0xFFFFFFFF, 0xFFAA0000));
            undobutton.setVisibility(View.VISIBLE);
        } else {
            dissectbutton.getBackground().setColorFilter(null);
            undobutton.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    //
    // API callbacks defined in HumanBodyInterface
    //
    /**
     * API Callback - model load complete
     */
    public void onModelLoaded() {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                body.ui.setBackgroundColor(Color.RED, Color.YELLOW);
                ProgressBar progress = findViewById(R.id.progressBar1);
                progress.setVisibility(View.GONE);
                // build Chapter pager
                HKChapter[] chaptersarray = new HKChapter[body.timeline.chapterList.size()];
                int i = 0;
                for ( String chapterid : body.timeline.chapterList ) {
                    chaptersarray[i] = body.timeline.chapters.get(chapterid);
                    i++;
                }
                ChapterAdapter adapter = new ChapterAdapter(getSupportFragmentManager());
                adapter.setChapters(chaptersarray);
                chapterPager.setAdapter(adapter);
            }
        });

    }

    Handler runHandler = new Handler(Looper.getMainLooper());
    float t;

    void camTest(float t) {
        float factor = 0.2f * (float)Math.sin(t);
        body.camera.zoom( factor );
        t = t + 0.1f;
        final float f = t;
        Runnable go = new Runnable() {
            public void run() {
                camTest(f);
            }
        };
        runHandler.postDelayed(go, 30);

    }

    public void onSceneInit(String title) {
        System.out.println("got model title " + title);
    }

    public void onModelLoaded(String title) {

    }

    /**
     * API Callback - object selected
     *
     * @param objectID the internal ID of the object
     */
    public void onObjectSelected(String objectID) {
        System.out.println("you selected " + body.scene.objects.get(objectID));
        View paintmenu = (View)findViewById(R.id.paintmenu);
        if (paintmenu.getVisibility() == View.VISIBLE) {
            if (paintColor == null) {
                body.scene.uncolor(objectID);
            } else {
                body.scene.color(objectID, paintColor);
            }
        }
    }

    /**
     * API Callback - object deselected
     *
     * @param objectId the internal ID of the object
     */
    public void onObjectDeselected(String objectId) {
        System.out.println("object deselected " + objectId);
    }

    public void onObjectsShown(Map<String,Object> shown) {
        for (String objectId : shown.keySet()) {
            Boolean showMe = (Boolean)shown.get(objectId);
//            System.out.println("object shown " + objectId + " = " + showMe);
        }
    }

    public void onTimelineUpdated(HKTimeline timeline) {
//        System.out.println("timeline update " + timeline.currentTime  + "/" + timeline.duration);
    }

    public void onSceneRestore() {
        System.out.println("scene restored");
    }

    public void onXrayEnabled(Boolean isEnabled) {
        System.out.println("xray enabled: " + isEnabled);
    }

    public void onAnnotationCreated(String annotationId) {
//        System.out.println("annotation created " + annotationId);
    }

    public void onAnnotationDestroyed(String annotationId) {
//        System.out.println("annotation destroyed " + annotationId);
    }

    public void onAnnotationsShown(Boolean shown) {
//        System.out.println("annotations shown " + shown);
    }

    public void onAnnotationUpdated(HKAnnotation annotation) {
//        System.out.println("annotation " + annotation.annotationId + " is at " + annotation.canvasPosition[0] + "," + annotation.canvasPosition[1]);
    }

    public void onObjectColor(String objectId, HKColor color) {
        System.out.println("got color for " + objectId + " color " + color.tint.toString());
    }

    public void onSceneCapture(String captureString) {
        System.out.println("** got scene captured message " + captureString);
    }

    public void onCameraUpdated(HKCamera camera) {
//        System.out.println("got camera update callback pos " + camera.eye[0] + "," + camera.eye[1] + "," + camera.eye[2] + " look " + camera.look[0] + "," + camera.look[1] + "," + camera.look[2] + ", up " + camera.up[0] + "," + camera.up[1] + "," + camera.up[2] + ", zoom " + camera.zoomFactor);
    }
    /**
     * API Callback - chapter transition to referenced chapter
     *
     * @param chapterID String ID of the Chapter, used to look up the Chapter object in
     *                  HumanBody's public HashMap<String,Chapter> chapters
     */
    public void onChapterTransition(String chapterID) {
        HKChapter chap = body.timeline.chapters.get(chapterID);
        System.out.println("got chapter " + chap.title);
    }

    /**
     *  API Callback - screenshot received
     *
     * @param image Bitmap object containing the screenshot image
     */
    public void onScreenshot(Bitmap image) {
        System.out.println("got screenshot");
    }

    /**
     *  API Callback - animation ended signal to reset animation UI
     */
    public void onAnimationComplete() {
    }

    public void onObjectPicked(String objectId, double[] position) {
        System.out.println("pick callback " + objectId + " at position " + position[0] + "," + position[1] + "," + position[2] );
    }

}

File: HumanKotlinApp/app/src/main/java/com/biodigital/kotlinapp/HumanActivity.kt
package com.biodigital.kotlinapp


import android.animation.LayoutTransition
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.LightingColorFilter
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager.widget.ViewPager
import com.biodigital.humansdk.*
import com.biodigital.kotlinapp.databinding.ActivityHumanBinding
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.MutableMap
import kotlin.collections.set
import kotlin.collections.toTypedArray

class HumanActivity : AppCompatActivity(), HKHumanInterface {

    private var xraymode = false
    private var isolatemode = false
    private var dissectmode = false
    private var paintmode = false

    private var humanbody: HKHuman? = null

    internal var paintColor : HKColor? = null

    internal var redColor = HKColor()
    internal var greenColor = HKColor()
    internal var blueColor = HKColor()
    internal var yellowColor = HKColor()

    internal var expanded = false
    private lateinit var binding: ActivityHumanBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityHumanBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        supportActionBar?.setDisplayHomeAsUpEnabled(true);

        val uiAll = false

        val modelId = intent.getStringExtra(MODEL_MESSAGE)
        System.out.println("load model $modelId")

        val uimap = HashMap<HumanUIOptions, Boolean>()
        uimap[HumanUIOptions.all] = uiAll

        humanbody = HKHuman(binding.humanview, uimap)

        humanbody!!.setInterface(this)

        humanbody!!.load(modelId)

        // turn the built in UI on, or off to show native UI elements
        val uiVisibility = if (uiAll) View.INVISIBLE else View.VISIBLE
        binding.homebutton.visibility = uiVisibility
        binding.resetbutton.visibility = uiVisibility
        binding.dissectbutton.visibility = uiVisibility
        binding.undobutton.visibility = View.INVISIBLE
        binding.xraybutton.visibility = uiVisibility
        binding.isolatebutton.visibility = uiVisibility
        binding.sharebutton.visibility = uiVisibility
        binding.allpaintstuff.visibility = uiVisibility
        binding.paintmenu.visibility = View.INVISIBLE
        binding.humanChapterPager.visibility = uiVisibility

        binding.homebutton.setOnClickListener { finish() }

        binding.resetbutton.setOnClickListener {
            humanbody!!.scene.reset()
            humanbody!!.camera.reset()
            xraymode = false
            isolatemode = false
            dissectmode = false
            binding.xraybutton.background.colorFilter = null
            binding.dissectbutton.background.colorFilter = null
            binding.isolatebutton.background.colorFilter = null
            binding.undobutton.visibility = View.INVISIBLE
        }

        binding.dissectbutton.setOnClickListener {
            if (paintmode) {
                binding.paintbutton.callOnClick();
            }
            dissectmode = !dissectmode
            humanbody!!.scene.dissect(dissectmode)
            binding.dissectbutton.isSelected = dissectmode
            if (dissectmode) {
                binding.dissectbutton.background.colorFilter = LightingColorFilter(-0x1, -0x560000)
                binding.undobutton.visibility = View.VISIBLE
            } else {
                binding.dissectbutton.background.colorFilter = null
                binding.undobutton.visibility = View.INVISIBLE
            }
        }

        binding.undobutton.setOnClickListener { humanbody!!.scene.undo() }

        binding.xraybutton.setOnClickListener {
            xraymode = !xraymode
            humanbody!!.scene.xray(xraymode)
            if (xraymode) {
                binding.xraybutton.background.colorFilter = LightingColorFilter(-0x1, -0x560000)
            } else {
                binding.xraybutton.background.colorFilter = null
            }
            if (dissectmode) {
                humanbody!!.scene.dissect(true)
            }
        }

        binding.isolatebutton.setOnClickListener {
            isolatemode = !isolatemode
            humanbody!!.scene.isolate(isolatemode)
            //                if (isolatemode) {
            //                    isolatebutton.getBackground().setColorFilter(new LightingColorFilter(0xFFFFFFFF, 0xFFAA0000));
            //                } else {
            //                    isolatebutton.getBackground().setColorFilter(null);
            //                }
            if (dissectmode) {
                humanbody!!.scene.dissect(true)
            }
        }

        binding.sharebutton.setOnClickListener { humanbody!!.scene.share() }

        binding.paintbutton.setOnClickListener {
            if (dissectmode) {
                binding.dissectbutton.callOnClick();
            }
            paintmode = !paintmode
            if (paintmode) {
                binding.paintmenu.visibility = View.VISIBLE
                humanbody!!.scene.disableHighlight()
                binding.paintbutton.background.colorFilter = LightingColorFilter(-0x1, -0x560000)
            } else {
                binding.paintmenu.visibility = View.INVISIBLE
                humanbody!!.scene.enableHighlight()
                binding.paintbutton.background.colorFilter = null
            }
        }

        redColor.tint = doubleArrayOf(1.0,0.0,0.0)
        greenColor.tint = doubleArrayOf(0.0,1.0,0.0)
        greenColor.saturation = 0.5;
        blueColor.tint = doubleArrayOf(0.0,0.0,1.0)
        blueColor.opacity = 0.66;
        yellowColor.tint = doubleArrayOf(1.0,1.0,0.0)

        binding.redbutton.setOnClickListener {
            paintColor = redColor
            binding.paintmenu.setBackgroundColor(Color.RED)
        }

        binding.greenbutton.setOnClickListener {
            paintColor = greenColor
            binding.paintmenu.setBackgroundColor(Color.GREEN)
        }

        binding.bluebutton.setOnClickListener {
            paintColor = blueColor
            binding.paintmenu.setBackgroundColor(Color.BLUE)
        }

        binding.yellowbutton.setOnClickListener {
            paintColor = yellowColor
            binding.paintmenu.setBackgroundColor(Color.YELLOW)
        }

        binding.undopaintbutton.setOnClickListener {
            paintColor = null
            binding.paintmenu.setBackgroundColor(Color.TRANSPARENT)
        }

        binding.humanChapterPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

            }

            override fun onPageSelected(position: Int) {
                if (position == humanbody!!.timeline.currentChapter.index) {
                    println("already at chapter $position")
                    return
                }
                if (position > humanbody!!.timeline.currentChapter.index) {
                    humanbody!!.timeline.nextChapter()
                } else {
                    humanbody!!.timeline.prevChapter()
                }
            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })

    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    fun handleChapterClick() {
        val scale = applicationContext.resources.displayMetrics.density
        (binding.category as ViewGroup).layoutTransition.enableTransitionType(LayoutTransition.CHANGING)
        if (expanded) {
            val px = (50 * scale + 0.5f).toInt()
            binding.category.layoutParams.height = px
        } else {
            val px = (160 * scale + 0.5f).toInt()
            binding.category.layoutParams.height = px
        }
        binding.category.requestLayout()
        expanded = !expanded
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    //
    // API callbacks defined in HumanBodyInterface
    //
    /**
     * API Callback - model load complete
     */
    override fun onModelLoaded(title: String) {
        println("MODEL LOADED CALLBACK $title")
        runOnUiThread {
            // build Chapter pager
            val chaptersarray = ArrayList<HKChapter>(humanbody!!.timeline.chapterList.size)
            for (chapterid in humanbody!!.timeline.chapterList) {
                var chapter = humanbody!!.timeline.chapters[chapterid]
                if (chapter != null) {
                    chaptersarray.add(chapter)
                }
            }
            val adapter = ChapterAdapter(supportFragmentManager)
            adapter.setChapters(chaptersarray.toTypedArray())
            binding.humanChapterPager.adapter = adapter
        }
    }

    override fun onModelLoadError(title: String) {
        println("load error")
    }

    /**
     * API Callback - object selected
     *
     * @param objectID the internal ID of the object
     */
    override fun onObjectSelected(objectId: String) {
        println("you picked " + humanbody!!.scene.objects[objectId]!!)
        if (binding.paintmenu.visibility == View.INVISIBLE) {
            return;
        }
        if (paintColor != null) {
            humanbody!!.scene.color(objectId, paintColor)
        } else {
            humanbody!!.scene.uncolor(objectId)
        }
    }

    override fun onObjectDeselected(objectId: String) {
    }

    /**
     * API Callback - chapter transition to referenced chapter
     *
     * @param chapterID String ID of the Chapter, used to look up the Chapter object in
     * HumanBody's public HashMap<String></String>,Chapter> chapters
     */
    override fun onChapterTransition(chapterID: String) {
        val chap = humanbody!!.timeline.chapters[chapterID]
        if (chap != null) {
            println("got chapter ${chap.title}")
        }
    }

    /**
     * API Callback - screenshot received
     *
     * @param image Bitmap object containing the screenshot image
     */
    override fun onScreenshot(image: Bitmap) {
        println("got screenshot")
    }

    /**
     * API Callback - animation ended signal to reset animation UI
     */
    override fun onAnimationComplete() {}

    override fun onAnnotationCreated(p0: String?) {
    }

    override fun onAnnotationDestroyed(p0: String?) {
    }

    override fun onAnnotationUpdated(p0: HKAnnotation?) {
    }

    override fun onAnnotationsShown(p0: Boolean?) {
    }

    override fun onCameraUpdated(p0: HKCamera?) {
    }

    override fun onObjectColor(p0: String?, p1: HKColor?) {
    }

    override fun onObjectPicked(p0: String?, p1: DoubleArray?) {
    }

    override fun onObjectsShown(p0: MutableMap<String, Any>?) {
    }

    override fun onSceneCapture(p0: String?) {
    }

    override fun onSceneInit(p0: String?) {
    }

    override fun onSceneRestore() {
    }

    override fun onTimelineUpdated(p0: HKTimeline?) {
    }

    override fun onXrayEnabled(p0: Boolean?) {
    }
}
